### PR TITLE
Add offline plugin

### LINKED ISSUE #19: Offline Plugin
This plugin would be a high level wrapper around  offline API.
More information and context in https://github.com/mapbox/mapbox-gl-native/issues/7289.

 - download a region
   - show notification while downloading
     - cancel button
     - customisation configuration
   - handle multiple requests
     - hook into onStartCommand
   - allow binding/unbinding to forward any errors?
   - handle connection state change events
 - bounds selection dialog
 - list offline regions
 - remove offline regions
 - change name offline region
 - update offline region content?
 - navigation?
 - searching for map data?
 - only download on wifi configuration
 - save maps only to SD card configuration?

### PR DESCRIPTION
**Pull request by tobrun**
WIP offline plugin, closes #19

### CODE REVIEW

**Comment by tobrun** [comment]
@zugaldia would you able to review current state? would love to merge this initial version and continue iterating on it with different issues and PRs. (will start working on ticketing out the issues).

**Comment by tobrun** [comment]
@zugaldia this PR is ready for review

File: plugins/build.gradle [hunk 1]
```diff
@@ -11,6 +11,7 @@ allprojects {
     repositories {
         jcenter()
         maven { url 'https://maven.google.com' }
+        maven { url "http://oss.sonatype.org/content/repositories/snapshots/" }
```
Comment (zugaldia) [inline plugins/build.gradle]: @tobrun We should make a note to remove this before final release (ticket?).

File: plugins/locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayer.java [hunk 2]
```diff
@@ -99,15 +99,24 @@ private void addSources() {
     // Create an empty feature collection
     FeatureCollection emptyFeature = FeatureCollection.fromFeatures(new Feature[] {});
 
-    GeoJsonSource locationSource = new GeoJsonSource(LocationLayerConstants.LOCATION_SOURCE, emptyFeature);
-    mapboxMap.addSource(locationSource);
+    if (mapboxMap.getSource(LocationLayerConstants.LOCATION_SOURCE) == null) {
```
Comment (zugaldia) [inline plugins/locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayer.java]: Are these changes required by 5.2? If so, 👍

File: plugins/offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/NotificationOptions.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+import static android.support.v4.app.NotificationCompat.Builder;
+
+public class NotificationOptions implements Parcelable {
+
+  @DrawableRes
+  private int smallIconRes;
+  private String returnActivity;
+  private String contentTitle = "Offline download";
+  private String contextText = "Downloading..";
```
Comment (zugaldia) [inline plugins/offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/NotificationOptions.java]: Missing period.

File: plugins/offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/OfflineDownloadOptions.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  }
+
+  public OfflineDownloadOptions withNotificationOptions(NotificationOptions notificationsOptions) {
+    this.notificationOptions = notificationsOptions;
+    return this;
+  }
+
+  // restrict visibility, only libs allowed to invoke building the OfflineDownload
+  OfflineDownload build() {
+    //TODO add validation
```
Comment (zugaldia) [inline plugins/offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/OfflineDownloadOptions.java]: Ticket?

File: plugins/offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/OfflineDownload.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+    progress = in.readInt();
+    metadata = new byte[in.readInt()];
+    in.readByteArray(metadata);
+  }
+
+  OfflineTilePyramidRegionDefinition getRegionDefinition() {
+    return definition;
+  }
+
+  byte[] getMetadata() {
```
Comment (zugaldia) [inline plugins/offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/OfflineDownload.java]: What if we exposed a higher level API for the metadata (can be ticketed separately). Instead of exposing the `byte[]` separately, we'd expose a POJO with, say, title and description, that we serialize/deserialize as part of this plugin?

Comment (zugaldia) [reply plugins/offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/OfflineDownload.java]: Just saw `OfflineUtils`, it simplifies things but I'd even go further than that and never expose that directly.

File: plugins/build.gradle (hunk 1 above)
Comment (tobrun) [reply plugins/build.gradle]: I consider above comment a great note for that ;)

File: plugins/locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayer.java (hunk 2 above)
Comment (tobrun) [reply plugins/locationlayer/src/main/java/com/mapbox/mapboxsdk/plugins/locationlayer/LocationLayer.java]: yes, it seems so, will fix this up in a seperate PR

File: plugins/offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/OfflineDownload.java (hunk 5 above)
Comment (tobrun) [reply plugins/offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/OfflineDownload.java]: Yeah, initially I went with String name and afterwards went back for byte[]. I'm liking the idea of  a POJO. To make that work, we will need an object that is Parceable as we need to be able to push that into a Bundle for Intent communication.

Comment (zugaldia) [reply plugins/offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/OfflineDownload.java]: @tobrun Or a small POJO easily convertible to JSON with Gson and use the String representation? Or would that potentially be too large for a Bundle?

File: plugins/offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/NotificationOptions.java [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+  public String getContentTitle() {
+    return contentTitle;
+  }
+
+  public NotificationOptions withContentText(String contextText) {
+    this.contextText = contextText;
+    return this;
+  }
+
+  public String getContextText() {
```
Comment (danesfeder) [inline plugins/offline/src/main/java/com/mapbox/mapboxsdk/plugins/offline/NotificationOptions.java]: @tobrun did you mean `getContentText()` here?

**Comment by tobrun** [comment]
@zugaldia I updated this PR to conform to the new project structure in https://github.com/mapbox/mapbox-plugins-android/pull/144. I would love to merge this to master so we can start building snapshots and iterate on it (will create some follow up issues so someone else on the team can pick them up)

**Comment by zugaldia** [comment]
> I would love to merge this to master so we can start building snapshots and iterate on it

Let's do it, just went ahead and approved the PR.

CI is complaining about the following:

```
* What went wrong:
A problem occurred evaluating script.
> java.lang.UnsupportedClassVersionError: org/sonarqube/gradle/SonarQubePlugin : Unsupported major.minor version 52.0
```

@cammace Do you know why? Is this because of the changes in https://github.com/mapbox/mapbox-plugins-android/pull/144?

**Comment by cammace** [comment]
The Bitrise script is still in this repo and ran correctly through all the same checks that circle ci would have. It's safe to ignore the fact that circleci is failing.
